Introduce 'implementationVersion' to JavaToolchainSpec

jjohannes opened the pull request:
This would allow you to select a precise JDK as follows:

```
java.toolchain {
    implementationVersion = "21.0.5"
}
```

### Context

For reasoning about this solution and more context:
https://github.com/gradle/gradle/issues/16628#issuecomment-2097540617

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md).
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team.
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective.
- ~Provide unit tests (under `<subproject>/src/test`) to verify logic.~
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes.
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`.
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

jjohannes committed:
Introduce 'implementationVersion' to JavaToolchainSpec

This would allow you to select a precise JDK as follows:

java.toolchain {
    implementationVersion = "21.0.5"
}

Signed-off-by: Jendrik Johannes <jendrik.johannes@gmail.com>

jvandort commented:
Thank you for your contribution!

Please add tests to cover your changes.  

***

I would like to see a test that actually goes and uses this property as part of a toolchain request

Related issues: https://github.com/gradle/gradle/issues/18896

jjohannes committed:
Add test for toolchain selection based on implementationVersion

Signed-off-by: Jendrik Johannes <jendrik.johannes@gmail.com>

jjohannes commented:
Thank you for considering this solution @jvandort. I added an integration test for the selection based on only the implementation version, which is how I would imagine this to be used.
I was looking for existing tests for different cases of toolchain selection but did not find dedicated tests on that topic. Hence I added a new test class. Let me know if you have ideas for other tests that would make sense in addition.

I can also add a sentence or two to the documentation, but I would only do that once the solution is accepted.

big-guy commented:
We are not against adding a way to specify/lock down the exact implementation version.  We're not going to be able to look at this before we finish the next major version (9.0).

big-guy commented:
@jjohannes this is on our list for 9.x, but it won't be able to make 9.3.

jjohannes committed:
Update 'since' to 9.6.0

Signed-off-by: Jendrik Johannes <jendrik@onepiece.software>

jjohannes committed:
Merge branch 'master' into java-toolchains-exact-version

jjohannes commented:
@big-guy @jvandort any chance this could be considered for the next release?

jjohannes committed:
Add missing Nullable annotation

Signed-off-by: Jendrik Johannes <jendrik@onepiece.software>
